src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderCursor.cs

```diff
@@ -576,45 +582,76 @@ public bool LastFieldIncludesNewLine(string line, bool startsInsideQuoted = fals
                             if (!startsInsideQuoted)
                                 ichCur++;
 
-                            for (; ; ichCur++)
+                            if (_escapeCharIsDoubleQuote)
                             {
```

Comment:
Is it possible to refactor these two if/else blocks into separate functions for the sake of readability? (Or even better merge them into a single function/codepath?

Reply:
It is possible to have this refactored into two functions or even into to one.

I didn't do this because the original writers of the Parser seemed to be very worried about efficiency and the way they addressed this is by avoiding creating / calling functions, and using bools such as the one I've used here for _escapeCharIsDoubleQuote.

So for example, we have the following code written by them, which can be obviously be refactored into a function, and perhaps even simplified to avoid the first 2 if branches, but it's stated that they didn't do it because it was "performance critical"
[https://github.com/dotnet/machinelearning/blob/c023271bf0233d4ebf6d55031ceb4233a965a4b7/src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderParser.cs#L1208-L1241](https://github.com/dotnet/machinelearning/blob/c023271bf0233d4ebf6d55031ceb4233a965a4b7/src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderParser.cs#L1208-L1241)

Following that pattern I thought that the fastest way for the code I am writing to run is to branch earlier with _escapeCharIsDoubleQuote only requiring to check it once... because if I refactor it to call functions or even worse if I merge into a single path that has to continuously check if _escapeCharIsDoubleQuote to work correctly, then it would have to take more steps and probably have a toll on performance. Also, personally I think that a merged path would be harder to read.

Do you still want me to try to refactor it? If so, I can try to come up with a refactored version for us to compare.

---
In reply to: [428386984](https://github.com/dotnet/machinelearning/pull/5147#discussion_r428386984) [](ancestors = 428386984)

Reply:
I couldn't find any benchmark tests for text loader in the code. Can you please look up and see if I have missed something?

---
In reply to: [428393614](https://github.com/dotnet/machinelearning/pull/5147#discussion_r428393614) [](ancestors = 428393614,428386984)

Reply:
I don't think there's any benchmark that only tests text loader. But I think it's a good idea to add one, so I'll do that.

In the meantime, I actually used the [FeaturizeTextBench](https://github.com/dotnet/machinelearning/blob/master/test/Microsoft.ML.Benchmarks/FeaturizeTextBench.cs) (which only loads a syntetic dataset and featurizes it) to run benchmarks on my previous readMultilines PR:
[https://github.com/dotnet/machinelearning/pull/5125#issuecomment-631094778](https://github.com/dotnet/machinelearning/pull/5125#issuecomment-631094778)

As I mention there, I'm not sure how to interpret the results, so I will address this tomorrow on scrum.

---
In reply to: [428398494](https://github.com/dotnet/machinelearning/pull/5147#discussion_r428398494) [](ancestors = 428398494,428393614,428386984)

Reply:
Comment in progress

---
In reply to: [428493474](https://github.com/dotnet/machinelearning/pull/5147#discussion_r428493474) [](ancestors = 428493474,428398494,428393614,428386984)